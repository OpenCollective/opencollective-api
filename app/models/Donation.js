
/*
 * Donation model
 * - this represents a commitment by a user to pay amount to a collective over a time period
 * - Note that there will only be one entry per subscription
 */

module.exports = function(Sequelize, DataTypes) {

  var Donation = Sequelize.define('Donation', {
    id: {
      type: DataTypes.INTEGER,
      primaryKey: true,
      autoIncrement: true
    },

    UserId: {
      type: DataTypes.INTEGER,
      references: 'Users',
      referencesKey: 'id',
      onDelete: 'SET NULL',
      onUpdate: 'CASCADE'
    },

    CollectiveId: {
      type: DataTypes.INTEGER,
      references: 'Groups',
      referencesKey: 'id',
      onDelete: 'SET NULL',
      onUpdate: 'CASCADE'
    },

    currency: {
      type: DataTypes.STRING,
      defaultValue: 'USD',
      set: function(val) {
        if (val && val.toUpperCase) {
          this.setDataValue('currency', val.toUpperCase());
        }
      }
    },

    amount: DataTypes.INTEGER,
    title: DataTypes.STRING,

    SubscriptionId: {
      type: DataTypes.INTEGER,
      references: 'Subscriptions',
      referencesKey: 'id',
      onDelete: 'SET NULL',
      onUpdate: 'CASCADE'
    },

    createdAt: {
      type: DataTypes.DATE,
      defaultValue: Sequelize.NOW
    },

    updatedAt: {
      type: DataTypes.DATE,
      defaultValue: Sequelize.NOW
    },

    deletedAt: {
      type: DataTypes.DATE
    }
  }, {
    paranoid: true,
    getterMethods: {
      // Info
      info() {
        return {
          id: this.id,
          UserId: this.UserId,
          CollectiveId: this.CollectiveId,
          currency: this.currency,
          amount: this.amount,
          amountFloat: this.amount / 100,
          title: this.title,
          SubscriptionId: this.SubscriptionId,
          createdAt: this.createdAt
        }
      }
    }
  });

  return Donation;
}
