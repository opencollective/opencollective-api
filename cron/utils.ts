import config from 'config';

/**
 * Heroku scheduler only has daily or hourly cron jobs, we only want to run
 * this script once per week on Monday (1). If the day is not Monday on production
 * we won't execute the script
 */
export function onlyExecuteInProdOnMondays() {
  const today = new Date();
  if (config.env === 'production' && today.getDay() !== 1) {
    console.log('OC_ENV is production and day is not Monday, script aborted!');
    process.exit(0);
  }
}
