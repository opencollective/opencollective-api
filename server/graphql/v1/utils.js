import debugLib from 'debug';
import { graphql } from 'graphql';

import { loaders } from '../loaders';

import schema from './schema';

const debug = debugLib('graphql');

export const makeRequest = (remoteUser, query) => {
  return {
    remoteUser,
    body: { query },
    loaders: loaders({ remoteUser }),
  };
};

export const graphqlQuery = async (query, variables, remoteUser) => {
  const prepare = () => {
    if (remoteUser) {
      remoteUser.rolesByCollectiveId = null; // force refetching the roles
      return remoteUser.populateRoles();
    } else {
      return Promise.resolve();
    }
  };

  if (process.env.DEBUG && process.env.DEBUG.match(/graphql/)) {
    debug('query', query);
    debug('variables', variables);
    debug('context', remoteUser);
  }

  return prepare().then(() =>
    graphql({
      schema,
      source: query,
      rootValue: null,
      contextValue: makeRequest(remoteUser, query),
      variableValues: variables,
    }),
  );
};
