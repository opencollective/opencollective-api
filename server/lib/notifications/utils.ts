import sanitizeHtml from 'sanitize-html';

import { sanitizerOptions as updateSanitizerOptions } from '../../models/Update';

const constructPreviewImageURL = (service: string, id: string) => {
  if (service === 'youtube' && id.match('[a-zA-Z0-9_-]{11}')) {
    return `https://img.youtube.com/vi/${id}/0.jpg`;
  } else if (service === 'anchorFm') {
    return `https://opencollective.com/static/images/anchor-fm-logo.png`;
  } else {
    return null;
  }
};

const parseServiceLink = (videoLink: string) => {
  const regexps = {
    youtube: new RegExp(
      '(?:https?://)?(?:www\\.)?youtu(?:\\.be/|be(-nocookie)?\\.com/\\S*(?:watch|embed)(?:(?:(?=/[^&\\s?]+(?!\\S))/)|(?:\\S*v=|v/)))([^&\\s?]+)',
      'i',
    ),
    anchorFm: /^(http|https)?:\/\/(www\.)?anchor\.fm\/([^/]+)(\/embed)?(\/episodes\/)?([^/]+)?\/?$/,
  };
  for (const service in regexps) {
    videoLink = videoLink.replace('/?showinfo=0', '');
    const matches = regexps[service].exec(videoLink);
    if (matches) {
      if (service === 'anchorFm') {
        const podcastName = matches[3];
        const episodeId = matches[6];
        const podcastUrl = `${podcastName}/embed`;
        return { service, id: episodeId ? `${podcastUrl}/episodes/${episodeId}` : podcastUrl };
      } else {
        return { service, id: matches[matches.length - 1] };
      }
    }
  }
  return {};
};

export const replaceVideosByImagePreviews = (html: string) => {
  const sanitizerOptions = {
    ...updateSanitizerOptions,
    transformTags: {
      ...updateSanitizerOptions.transformTags,
      iframe: (tagName, attribs) => {
        if (!attribs.src) {
          return '';
        }
        const { service, id } = parseServiceLink(attribs.src);
        const imgSrc = constructPreviewImageURL(service, id);
        if (imgSrc) {
          return {
            tagName: 'img',
            attribs: {
              src: imgSrc,
              alt: `${service} content`,
            },
          };
        } else {
          return '';
        }
      },
    },
  };
  return sanitizeHtml(html, sanitizerOptions);
};
