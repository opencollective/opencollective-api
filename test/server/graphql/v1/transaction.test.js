/**
 * Note: to update the snapshots run:
 * TZ=UTC CHAI_JEST_SNAPSHOT_UPDATE_ALL=true npx mocha test/server/graphql/v1/transaction.test.js
 */

import { expect } from 'chai';
import gql from 'fake-tag';
import { describe, it } from 'mocha';

import models from '../../../../server/models';
import * as store from '../../../stores';
import * as utils from '../../../utils';

describe('server/graphql/v1/transaction', () => {
  before(async () => {
    await utils.resetTestDB();
    // Given a host
    const { hostCollective } = await store.newHost('wwcode', 'USD', 5);
    // Given a collective
    const { collective } = await store.newCollectiveInHost('wwcodeaustin', 'USD', hostCollective, null, {
      isActive: true,
    });
    // And given the host has a stripe account
    await store.stripeConnectedAccount(hostCollective.id);
    // And given that we have 5 users making one purchase each
    const userNames = ['Craft Work', 'Geoff Frey', 'Holly Day', 'Max Point', 'Praxent'];
    // Not using Promise.all because I want the entries to be created
    // in sequence, not in parallel since stripeOneTimeDonation can't
    // patch the same object more than once at a time.
    for (let i = 0; i < userNames.length; i++) {
      const { user } = await store.newUser(userNames[i]);
      await store.stripeOneTimeDonation({
        remoteUser: user,
        collective,
        currency: 'USD',
        amount: 1000 * (i + 1),
        createdAt: new Date(2018, i, i, 0, 0, i),
      });
      await store.stripeOneTimeDonation({
        remoteUser: user,
        collective,
        currency: 'USD',
        amount: 1000 * (i + 1),
        createdAt: new Date(2017, i, i, 0, 0, i),
      });
    }
  });

  describe('return collective.transactions', () => {
    it('when given a collective slug (case insensitive)', async () => {
      const limit = 40;
      const collectiveQuery = gql`
        query Collective($slug: String, $limit: Int) {
          Collective(slug: $slug) {
            id
            slug
            transactions(limit: $limit) {
              id
              type
              createdByUser {
                id
                email
              }
              host {
                id
                slug
              }
              ... on Order {
                paymentMethod {
                  id
                  name
                }
                subscription {
                  id
                  interval
                }
              }
            }
          }
        }
      `;
      const result = await utils.graphqlQuery(collectiveQuery, {
        slug: 'WWCodeAustin',
        limit,
      });
      expect(result.data.Collective).to.exist;
      expect(result.data.Collective.transactions).to.have.length(20);
      expect(result).to.matchSnapshot();
    });
  });

  describe('return transactions', () => {
    it('returns one transaction by id', async () => {
      const transactionQuery = gql`
        query Transaction($id: Int) {
          Transaction(id: $id) {
            id
            type
            createdByUser {
              id
              email
            }
            host {
              id
              slug
            }
            ... on Order {
              paymentMethod {
                id
                name
              }
              subscription {
                id
                interval
              }
            }
          }
        }
      `;
      const result = await utils.graphqlQuery(transactionQuery, { id: 2 });
      expect(result).to.matchSnapshot();
    });

    it('returns one transaction by uuid', async () => {
      const transactionQuery = gql`
        query Transaction($uuid: String) {
          Transaction(uuid: $uuid) {
            id
            type
            createdByUser {
              id
              email
            }
            host {
              id
              slug
            }
            ... on Order {
              paymentMethod {
                id
                name
              }
              subscription {
                id
                interval
              }
            }
          }
        }
      `;
      const transaction = await models.Transaction.findOne({ where: { id: 1 } });
      const result = await utils.graphqlQuery(transactionQuery, {
        uuid: transaction.uuid,
      });
      expect(result).to.matchSnapshot();
    });

    it('with filter on type', async () => {
      const limit = 100;
      const allTransactionsQuery = gql`
        query AllTransactions($CollectiveId: Int!, $limit: Int, $offset: Int, $type: String) {
          allTransactions(CollectiveId: $CollectiveId, limit: $limit, offset: $offset, type: $type) {
            id
            type
          }
        }
      `;
      const result = await utils.graphqlQuery(allTransactionsQuery, {
        CollectiveId: 3,
        limit,
        type: 'CREDIT',
      });
      expect(result.data.allTransactions).to.have.length(10);
      expect(result).to.matchSnapshot();
    });

    it('with dateFrom', async () => {
      // Given the followin query
      const allTransactionsQuery = gql`
        query AllTransactions(
          $CollectiveId: Int!
          $limit: Int
          $offset: Int
          $type: String
          $dateFrom: String
          $dateTo: String
        ) {
          allTransactions(
            CollectiveId: $CollectiveId
            limit: $limit
            offset: $offset
            type: $type
            dateFrom: $dateFrom
            dateTo: $dateTo
          ) {
            id
            createdAt
          }
        }
      `;

      // When the query is executed with the parameter `dateFrom`
      const result = await utils.graphqlQuery(allTransactionsQuery, {
        CollectiveId: 3,
        dateFrom: '2017-10-01',
      });
      expect(result.data.allTransactions).to.have.length(10);
      expect(result).to.matchSnapshot();
    });

    it('with dateTo', async () => {
      // Given the followin query
      const allTransactionsQuery = gql`
        query AllTransactions(
          $CollectiveId: Int!
          $limit: Int
          $offset: Int
          $type: String
          $dateFrom: String
          $dateTo: String
        ) {
          allTransactions(
            CollectiveId: $CollectiveId
            limit: $limit
            offset: $offset
            type: $type
            dateFrom: $dateFrom
            dateTo: $dateTo
          ) {
            id
            createdAt
          }
        }
      `;

      // When the query is executed with the parameter `dateTo`
      const result = await utils.graphqlQuery(allTransactionsQuery, {
        CollectiveId: 3,
        dateTo: '2017-10-01',
      });
      expect(result.data.allTransactions).to.have.length(10);
      expect(result).to.matchSnapshot();
    });

    it('with pagination', async () => {
      const limit = 20;
      const offset = 0;
      const allTransactionsQuery = gql`
        query AllTransactions($CollectiveId: Int!, $limit: Int, $offset: Int) {
          allTransactions(CollectiveId: $CollectiveId, limit: $limit, offset: $offset) {
            id
            type
            kind
            createdByUser {
              id
              email
            }
            fromCollective {
              id
              slug
            }
            collective {
              id
              slug
            }
            host {
              id
              slug
            }
            ... on Order {
              paymentMethod {
                id
                name
              }
              subscription {
                id
                interval
              }
            }
          }
        }
      `;
      const result = await utils.graphqlQuery(allTransactionsQuery, {
        CollectiveId: 3,
        limit,
        offset,
      });
      expect(result).to.matchSnapshot();
    });
  });
});
