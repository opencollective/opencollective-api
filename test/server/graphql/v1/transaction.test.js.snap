// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`server/graphql/v1/transaction return collective.transactions when given a collective slug (case insensitive) 1`] = `
Object {
  "data": Object {
    "Collective": Object {
      "id": 3,
      "slug": "wwcodeaustin",
      "transactions": Array [
        Object {
          "createdByUser": Object {
            "email": null,
            "id": 7,
          },
          "host": Object {
            "id": 2,
            "slug": "wwcode",
          },
          "id": 40,
          "paymentMethod": Object {
            "id": 13,
            "name": null,
          },
          "subscription": null,
          "type": "CREDIT",
        },
        Object {
          "createdByUser": null,
          "host": Object {
            "id": 2,
            "slug": "wwcode",
          },
          "id": 37,
          "type": "DEBIT",
        },
        Object {
          "createdByUser": Object {
            "email": null,
            "id": 7,
          },
          "host": Object {
            "id": 2,
            "slug": "wwcode",
          },
          "id": 36,
          "paymentMethod": Object {
            "id": 12,
            "name": null,
          },
          "subscription": null,
          "type": "CREDIT",
        },
        Object {
          "createdByUser": null,
          "host": Object {
            "id": 2,
            "slug": "wwcode",
          },
          "id": 33,
          "type": "DEBIT",
        },
        Object {
          "createdByUser": Object {
            "email": null,
            "id": 6,
          },
          "host": Object {
            "id": 2,
            "slug": "wwcode",
          },
          "id": 32,
          "paymentMethod": Object {
            "id": 11,
            "name": null,
          },
          "subscription": null,
          "type": "CREDIT",
        },
        Object {
          "createdByUser": null,
          "host": Object {
            "id": 2,
            "slug": "wwcode",
          },
          "id": 29,
          "type": "DEBIT",
        },
        Object {
          "createdByUser": Object {
            "email": null,
            "id": 6,
          },
          "host": Object {
            "id": 2,
            "slug": "wwcode",
          },
          "id": 28,
          "paymentMethod": Object {
            "id": 10,
            "name": null,
          },
          "subscription": null,
          "type": "CREDIT",
        },
        Object {
          "createdByUser": null,
          "host": Object {
            "id": 2,
            "slug": "wwcode",
          },
          "id": 25,
          "type": "DEBIT",
        },
        Object {
          "createdByUser": Object {
            "email": null,
            "id": 5,
          },
          "host": Object {
            "id": 2,
            "slug": "wwcode",
          },
          "id": 24,
          "paymentMethod": Object {
            "id": 9,
            "name": null,
          },
          "subscription": null,
          "type": "CREDIT",
        },
        Object {
          "createdByUser": null,
          "host": Object {
            "id": 2,
            "slug": "wwcode",
          },
          "id": 21,
          "type": "DEBIT",
        },
        Object {
          "createdByUser": Object {
            "email": null,
            "id": 5,
          },
          "host": Object {
            "id": 2,
            "slug": "wwcode",
          },
          "id": 20,
          "paymentMethod": Object {
            "id": 8,
            "name": null,
          },
          "subscription": null,
          "type": "CREDIT",
        },
        Object {
          "createdByUser": null,
          "host": Object {
            "id": 2,
            "slug": "wwcode",
          },
          "id": 17,
          "type": "DEBIT",
        },
        Object {
          "createdByUser": Object {
            "email": null,
            "id": 4,
          },
          "host": Object {
            "id": 2,
            "slug": "wwcode",
          },
          "id": 16,
          "paymentMethod": Object {
            "id": 7,
            "name": null,
          },
          "subscription": null,
          "type": "CREDIT",
        },
        Object {
          "createdByUser": null,
          "host": Object {
            "id": 2,
            "slug": "wwcode",
          },
          "id": 13,
          "type": "DEBIT",
        },
        Object {
          "createdByUser": Object {
            "email": null,
            "id": 4,
          },
          "host": Object {
            "id": 2,
            "slug": "wwcode",
          },
          "id": 12,
          "paymentMethod": Object {
            "id": 6,
            "name": null,
          },
          "subscription": null,
          "type": "CREDIT",
        },
        Object {
          "createdByUser": null,
          "host": Object {
            "id": 2,
            "slug": "wwcode",
          },
          "id": 9,
          "type": "DEBIT",
        },
        Object {
          "createdByUser": Object {
            "email": null,
            "id": 3,
          },
          "host": Object {
            "id": 2,
            "slug": "wwcode",
          },
          "id": 8,
          "paymentMethod": Object {
            "id": 5,
            "name": null,
          },
          "subscription": null,
          "type": "CREDIT",
        },
        Object {
          "createdByUser": null,
          "host": Object {
            "id": 2,
            "slug": "wwcode",
          },
          "id": 5,
          "type": "DEBIT",
        },
        Object {
          "createdByUser": Object {
            "email": null,
            "id": 3,
          },
          "host": Object {
            "id": 2,
            "slug": "wwcode",
          },
          "id": 4,
          "paymentMethod": Object {
            "id": 4,
            "name": null,
          },
          "subscription": null,
          "type": "CREDIT",
        },
        Object {
          "createdByUser": null,
          "host": Object {
            "id": 2,
            "slug": "wwcode",
          },
          "id": 1,
          "type": "DEBIT",
        },
      ],
    },
  },
}
`;

exports[`server/graphql/v1/transaction return transactions returns one transaction by id 1`] = `
Object {
  "data": Object {
    "Transaction": Object {
      "createdByUser": null,
      "host": Object {
        "id": 2,
        "slug": "wwcode",
      },
      "id": 2,
      "paymentMethod": null,
      "subscription": null,
      "type": "CREDIT",
    },
  },
}
`;

exports[`server/graphql/v1/transaction return transactions returns one transaction by uuid 1`] = `
Object {
  "data": Object {
    "Transaction": Object {
      "createdByUser": null,
      "host": Object {
        "id": 2,
        "slug": "wwcode",
      },
      "id": 1,
      "type": "DEBIT",
    },
  },
}
`;

exports[`server/graphql/v1/transaction return transactions with dateFrom 1`] = `
Object {
  "data": Object {
    "allTransactions": Array [
      Object {
        "createdAt": "Fri May 04 2018 00:00:04 GMT+0000 (Coordinated Universal Time)",
        "id": 36,
      },
      Object {
        "createdAt": "Fri May 04 2018 00:00:04 GMT+0000 (Coordinated Universal Time)",
        "id": 33,
      },
      Object {
        "createdAt": "Tue Apr 03 2018 00:00:03 GMT+0000 (Coordinated Universal Time)",
        "id": 28,
      },
      Object {
        "createdAt": "Tue Apr 03 2018 00:00:03 GMT+0000 (Coordinated Universal Time)",
        "id": 25,
      },
      Object {
        "createdAt": "Fri Mar 02 2018 00:00:02 GMT+0000 (Coordinated Universal Time)",
        "id": 20,
      },
      Object {
        "createdAt": "Fri Mar 02 2018 00:00:02 GMT+0000 (Coordinated Universal Time)",
        "id": 17,
      },
      Object {
        "createdAt": "Thu Feb 01 2018 00:00:01 GMT+0000 (Coordinated Universal Time)",
        "id": 12,
      },
      Object {
        "createdAt": "Thu Feb 01 2018 00:00:01 GMT+0000 (Coordinated Universal Time)",
        "id": 9,
      },
      Object {
        "createdAt": "Sun Dec 31 2017 00:00:00 GMT+0000 (Coordinated Universal Time)",
        "id": 4,
      },
      Object {
        "createdAt": "Sun Dec 31 2017 00:00:00 GMT+0000 (Coordinated Universal Time)",
        "id": 1,
      },
    ],
  },
}
`;

exports[`server/graphql/v1/transaction return transactions with dateTo 1`] = `
Object {
  "data": Object {
    "allTransactions": Array [
      Object {
        "createdAt": "Thu May 04 2017 00:00:04 GMT+0000 (Coordinated Universal Time)",
        "id": 40,
      },
      Object {
        "createdAt": "Thu May 04 2017 00:00:04 GMT+0000 (Coordinated Universal Time)",
        "id": 37,
      },
      Object {
        "createdAt": "Mon Apr 03 2017 00:00:03 GMT+0000 (Coordinated Universal Time)",
        "id": 32,
      },
      Object {
        "createdAt": "Mon Apr 03 2017 00:00:03 GMT+0000 (Coordinated Universal Time)",
        "id": 29,
      },
      Object {
        "createdAt": "Thu Mar 02 2017 00:00:02 GMT+0000 (Coordinated Universal Time)",
        "id": 24,
      },
      Object {
        "createdAt": "Thu Mar 02 2017 00:00:02 GMT+0000 (Coordinated Universal Time)",
        "id": 21,
      },
      Object {
        "createdAt": "Wed Feb 01 2017 00:00:01 GMT+0000 (Coordinated Universal Time)",
        "id": 16,
      },
      Object {
        "createdAt": "Wed Feb 01 2017 00:00:01 GMT+0000 (Coordinated Universal Time)",
        "id": 13,
      },
      Object {
        "createdAt": "Sat Dec 31 2016 00:00:00 GMT+0000 (Coordinated Universal Time)",
        "id": 8,
      },
      Object {
        "createdAt": "Sat Dec 31 2016 00:00:00 GMT+0000 (Coordinated Universal Time)",
        "id": 5,
      },
    ],
  },
}
`;

exports[`server/graphql/v1/transaction return transactions with filter on type 1`] = `
Object {
  "data": Object {
    "allTransactions": Array [
      Object {
        "id": 36,
        "type": "CREDIT",
      },
      Object {
        "id": 28,
        "type": "CREDIT",
      },
      Object {
        "id": 20,
        "type": "CREDIT",
      },
      Object {
        "id": 12,
        "type": "CREDIT",
      },
      Object {
        "id": 4,
        "type": "CREDIT",
      },
      Object {
        "id": 40,
        "type": "CREDIT",
      },
      Object {
        "id": 32,
        "type": "CREDIT",
      },
      Object {
        "id": 24,
        "type": "CREDIT",
      },
      Object {
        "id": 16,
        "type": "CREDIT",
      },
      Object {
        "id": 8,
        "type": "CREDIT",
      },
    ],
  },
}
`;

exports[`server/graphql/v1/transaction return transactions with pagination 1`] = `
Object {
  "data": Object {
    "allTransactions": Array [
      Object {
        "collective": Object {
          "id": 3,
          "slug": "wwcodeaustin",
        },
        "createdByUser": Object {
          "email": null,
          "id": 7,
        },
        "fromCollective": Object {
          "id": 9,
          "slug": "praxent",
        },
        "host": Object {
          "id": 2,
          "slug": "wwcode",
        },
        "id": 36,
        "kind": "CONTRIBUTION",
        "paymentMethod": Object {
          "id": 12,
          "name": null,
        },
        "subscription": null,
        "type": "CREDIT",
      },
      Object {
        "collective": Object {
          "id": 3,
          "slug": "wwcodeaustin",
        },
        "createdByUser": null,
        "fromCollective": Object {
          "id": 2,
          "slug": "wwcode",
        },
        "host": Object {
          "id": 2,
          "slug": "wwcode",
        },
        "id": 33,
        "kind": "HOST_FEE",
        "type": "DEBIT",
      },
      Object {
        "collective": Object {
          "id": 3,
          "slug": "wwcodeaustin",
        },
        "createdByUser": Object {
          "email": null,
          "id": 6,
        },
        "fromCollective": Object {
          "id": 8,
          "slug": "max-point",
        },
        "host": Object {
          "id": 2,
          "slug": "wwcode",
        },
        "id": 28,
        "kind": "CONTRIBUTION",
        "paymentMethod": Object {
          "id": 10,
          "name": null,
        },
        "subscription": null,
        "type": "CREDIT",
      },
      Object {
        "collective": Object {
          "id": 3,
          "slug": "wwcodeaustin",
        },
        "createdByUser": null,
        "fromCollective": Object {
          "id": 2,
          "slug": "wwcode",
        },
        "host": Object {
          "id": 2,
          "slug": "wwcode",
        },
        "id": 25,
        "kind": "HOST_FEE",
        "type": "DEBIT",
      },
      Object {
        "collective": Object {
          "id": 3,
          "slug": "wwcodeaustin",
        },
        "createdByUser": Object {
          "email": null,
          "id": 5,
        },
        "fromCollective": Object {
          "id": 7,
          "slug": "holly-day",
        },
        "host": Object {
          "id": 2,
          "slug": "wwcode",
        },
        "id": 20,
        "kind": "CONTRIBUTION",
        "paymentMethod": Object {
          "id": 8,
          "name": null,
        },
        "subscription": null,
        "type": "CREDIT",
      },
      Object {
        "collective": Object {
          "id": 3,
          "slug": "wwcodeaustin",
        },
        "createdByUser": null,
        "fromCollective": Object {
          "id": 2,
          "slug": "wwcode",
        },
        "host": Object {
          "id": 2,
          "slug": "wwcode",
        },
        "id": 17,
        "kind": "HOST_FEE",
        "type": "DEBIT",
      },
      Object {
        "collective": Object {
          "id": 3,
          "slug": "wwcodeaustin",
        },
        "createdByUser": Object {
          "email": null,
          "id": 4,
        },
        "fromCollective": Object {
          "id": 6,
          "slug": "geoff-frey",
        },
        "host": Object {
          "id": 2,
          "slug": "wwcode",
        },
        "id": 12,
        "kind": "CONTRIBUTION",
        "paymentMethod": Object {
          "id": 6,
          "name": null,
        },
        "subscription": null,
        "type": "CREDIT",
      },
      Object {
        "collective": Object {
          "id": 3,
          "slug": "wwcodeaustin",
        },
        "createdByUser": null,
        "fromCollective": Object {
          "id": 2,
          "slug": "wwcode",
        },
        "host": Object {
          "id": 2,
          "slug": "wwcode",
        },
        "id": 9,
        "kind": "HOST_FEE",
        "type": "DEBIT",
      },
      Object {
        "collective": Object {
          "id": 3,
          "slug": "wwcodeaustin",
        },
        "createdByUser": Object {
          "email": null,
          "id": 3,
        },
        "fromCollective": Object {
          "id": 5,
          "slug": "craft-work",
        },
        "host": Object {
          "id": 2,
          "slug": "wwcode",
        },
        "id": 4,
        "kind": "CONTRIBUTION",
        "paymentMethod": Object {
          "id": 4,
          "name": null,
        },
        "subscription": null,
        "type": "CREDIT",
      },
      Object {
        "collective": Object {
          "id": 3,
          "slug": "wwcodeaustin",
        },
        "createdByUser": null,
        "fromCollective": Object {
          "id": 2,
          "slug": "wwcode",
        },
        "host": Object {
          "id": 2,
          "slug": "wwcode",
        },
        "id": 1,
        "kind": "HOST_FEE",
        "type": "DEBIT",
      },
      Object {
        "collective": Object {
          "id": 3,
          "slug": "wwcodeaustin",
        },
        "createdByUser": Object {
          "email": null,
          "id": 7,
        },
        "fromCollective": Object {
          "id": 9,
          "slug": "praxent",
        },
        "host": Object {
          "id": 2,
          "slug": "wwcode",
        },
        "id": 40,
        "kind": "CONTRIBUTION",
        "paymentMethod": Object {
          "id": 13,
          "name": null,
        },
        "subscription": null,
        "type": "CREDIT",
      },
      Object {
        "collective": Object {
          "id": 3,
          "slug": "wwcodeaustin",
        },
        "createdByUser": null,
        "fromCollective": Object {
          "id": 2,
          "slug": "wwcode",
        },
        "host": Object {
          "id": 2,
          "slug": "wwcode",
        },
        "id": 37,
        "kind": "HOST_FEE",
        "type": "DEBIT",
      },
      Object {
        "collective": Object {
          "id": 3,
          "slug": "wwcodeaustin",
        },
        "createdByUser": Object {
          "email": null,
          "id": 6,
        },
        "fromCollective": Object {
          "id": 8,
          "slug": "max-point",
        },
        "host": Object {
          "id": 2,
          "slug": "wwcode",
        },
        "id": 32,
        "kind": "CONTRIBUTION",
        "paymentMethod": Object {
          "id": 11,
          "name": null,
        },
        "subscription": null,
        "type": "CREDIT",
      },
      Object {
        "collective": Object {
          "id": 3,
          "slug": "wwcodeaustin",
        },
        "createdByUser": null,
        "fromCollective": Object {
          "id": 2,
          "slug": "wwcode",
        },
        "host": Object {
          "id": 2,
          "slug": "wwcode",
        },
        "id": 29,
        "kind": "HOST_FEE",
        "type": "DEBIT",
      },
      Object {
        "collective": Object {
          "id": 3,
          "slug": "wwcodeaustin",
        },
        "createdByUser": Object {
          "email": null,
          "id": 5,
        },
        "fromCollective": Object {
          "id": 7,
          "slug": "holly-day",
        },
        "host": Object {
          "id": 2,
          "slug": "wwcode",
        },
        "id": 24,
        "kind": "CONTRIBUTION",
        "paymentMethod": Object {
          "id": 9,
          "name": null,
        },
        "subscription": null,
        "type": "CREDIT",
      },
      Object {
        "collective": Object {
          "id": 3,
          "slug": "wwcodeaustin",
        },
        "createdByUser": null,
        "fromCollective": Object {
          "id": 2,
          "slug": "wwcode",
        },
        "host": Object {
          "id": 2,
          "slug": "wwcode",
        },
        "id": 21,
        "kind": "HOST_FEE",
        "type": "DEBIT",
      },
      Object {
        "collective": Object {
          "id": 3,
          "slug": "wwcodeaustin",
        },
        "createdByUser": Object {
          "email": null,
          "id": 4,
        },
        "fromCollective": Object {
          "id": 6,
          "slug": "geoff-frey",
        },
        "host": Object {
          "id": 2,
          "slug": "wwcode",
        },
        "id": 16,
        "kind": "CONTRIBUTION",
        "paymentMethod": Object {
          "id": 7,
          "name": null,
        },
        "subscription": null,
        "type": "CREDIT",
      },
      Object {
        "collective": Object {
          "id": 3,
          "slug": "wwcodeaustin",
        },
        "createdByUser": null,
        "fromCollective": Object {
          "id": 2,
          "slug": "wwcode",
        },
        "host": Object {
          "id": 2,
          "slug": "wwcode",
        },
        "id": 13,
        "kind": "HOST_FEE",
        "type": "DEBIT",
      },
      Object {
        "collective": Object {
          "id": 3,
          "slug": "wwcodeaustin",
        },
        "createdByUser": Object {
          "email": null,
          "id": 3,
        },
        "fromCollective": Object {
          "id": 5,
          "slug": "craft-work",
        },
        "host": Object {
          "id": 2,
          "slug": "wwcode",
        },
        "id": 8,
        "kind": "CONTRIBUTION",
        "paymentMethod": Object {
          "id": 5,
          "name": null,
        },
        "subscription": null,
        "type": "CREDIT",
      },
      Object {
        "collective": Object {
          "id": 3,
          "slug": "wwcodeaustin",
        },
        "createdByUser": null,
        "fromCollective": Object {
          "id": 2,
          "slug": "wwcode",
        },
        "host": Object {
          "id": 2,
          "slug": "wwcode",
        },
        "id": 5,
        "kind": "HOST_FEE",
        "type": "DEBIT",
      },
    ],
  },
}
`;
